runtime/ast/unmarshal.go

```diff
@@ -0,0 +1,901 @@
+		Expression: ExpressionUnmarshaler{&a.Expression},
+	})
+}
+
+func (a *Program) UnmarshalJSON(b []byte) error {
+	type Alias Program
+	return json.Unmarshal(b, &struct {
+		*Alias
+		Declarations *[]Declaration
+	}{
+		Alias:        (*Alias)(a),
+		Declarations: &a.declarations,
+	})
+}
+
+func (a *ReferenceType) UnmarshalJSON(b []byte) error {
+	type Alias ReferenceType
+	return json.Unmarshal(b, &struct {
+		*Alias
+		ReferencedType TypeUnmarshaler
+		StartPos       *Position
+	}{
+		Alias:          (*Alias)(a),
+		ReferencedType: TypeUnmarshaler{&a.Type},
+		StartPos:       &a.StartPos,
+	})
+}
+
+func (a *RestrictedType) UnmarshalJSON(b []byte) error {
+	type Alias RestrictedType
+	return json.Unmarshal(b, &struct {
+		*Alias
+		RestrictedType TypeUnmarshaler
+	}{
+		Alias:          (*Alias)(a),
+		RestrictedType: TypeUnmarshaler{&a.Type},
+	})
+}
+
+func (a *ReturnStatement) UnmarshalJSON(b []byte) error {
+	type Alias ReturnStatement
+	return json.Unmarshal(b, &struct {
+		*Alias
+		Expression ExpressionUnmarshaler
+	}{
+		Alias:      (*Alias)(a),
+		Expression: ExpressionUnmarshaler{&a.Expression},
+	})
+}
+
+func (a *SwapStatement) UnmarshalJSON(b []byte) error {
+	type Alias SwapStatement
+	return json.Unmarshal(b, &struct {
+		*Alias
+		Left  ExpressionUnmarshaler
+		Right ExpressionUnmarshaler
+	}{
+		Alias: (*Alias)(a),
+		Left:  ExpressionUnmarshaler{&a.Left},
+		Right: ExpressionUnmarshaler{&a.Right},
+	})
+}
+
+func (a *SwitchStatement) UnmarshalJSON(b []byte) error {
+	type Alias SwitchStatement
+	return json.Unmarshal(b, &struct {
+		*Alias
+		Expression ExpressionUnmarshaler
+	}{
+		Alias:      (*Alias)(a),
+		Expression: ExpressionUnmarshaler{&a.Expression},
+	})
+}
+
+func (a *Transfer) UnmarshalJSON(b []byte) error {
+	type Alias Transfer
+	return json.Unmarshal(b, &struct {
+		*Alias
+		StartPos *Position
+	}{
+		Alias:    (*Alias)(a),
+		StartPos: &a.Pos,
+	})
+}
+
+func (a *TransferOperation) UnmarshalJSON(b []byte) error {
+	if op, ok := transferOperationMap[strings.Trim(string(b), `"`)]; ok {
+		*a = op
+		return nil
+	}
+	return fmt.Errorf("unknown TransferOperation")
+}
+
+func (a *TypeAnnotation) UnmarshalJSON(b []byte) error {
+	type Alias TypeAnnotation
+	return json.Unmarshal(b, &struct {
+		*Alias
+		AnnotatedType TypeUnmarshaler
+		StartPos      *Position
+	}{
+		Alias:         (*Alias)(a),
+		AnnotatedType: TypeUnmarshaler{&a.Type},
+		StartPos:      &a.StartPos,
+	})
+}
+
+func (a *VariableDeclaration) UnmarshalJSON(b []byte) error {
+	type Alias VariableDeclaration
+	return json.Unmarshal(b, &struct {
+		*Alias
+		Value       ExpressionUnmarshaler
+		SecondValue ExpressionUnmarshaler
+		StartPos    *Position
+	}{
+		Alias:       (*Alias)(a),
+		Value:       ExpressionUnmarshaler{&a.Value},
+		SecondValue: ExpressionUnmarshaler{&a.SecondValue},
+		StartPos:    &a.StartPos,
+	})
+}
+
+func (a *VariableKind) UnmarshalJSON(b []byte) error {
+	if vk, ok := variableKindMap[strings.Trim(string(b), `"`)]; ok {
```

Comment:
Instead of trimming the serialized string's quotes manually, potentially leaving quoted content, unmarshal as a string:

```go
var s string
if err := json.Unmarshal(b, &s); err != nil {
	return err
}
```